File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
while True:
    arr = list(str(input().rstrip()))
    if len(arr) == 1 and arr[0]=='.': # 만약 .이 나오면 종료
        break
    stack = []
    for i in arr:
        if i == '(' or i =='[':
            stack.append(i)
        elif i == ')':
            if len(stack)!=0 and stack[-1]=='(':
                stack.pop()
            else:
                stack.append(i)
                break
        elif i == ']':
            if len(stack)!=0 and stack[-1]=='[':
                stack.pop()
            else:
                stack.append(']')
                break
        else:
            continue

    if len(stack)!=0:
        print('no')
    else:
        print('yes')

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;
import java.util.stream.Collectors;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        while(true){
            Stack<String> stack = new Stack<>();
            String[] str = br.readLine().split("");

            // 종료 조건
            if(str.length==1 && str[0].equals(".")){
                break;
            }
            boolean flag = false;
            for (String s:str) {
                if(s.equals("(")||s.equals("[")) stack.push(s);
                else if(s.equals(")")&&!stack.isEmpty()){
                    if(stack.peek().equals("(")){
                        stack.pop();
                    }else{
                        flag = true;
                        break;
                    }
                }
                else if(s.equals("]")&&!stack.isEmpty()){
                    if(stack.peek().equals("[")){
                        stack.pop();
                    }else{
                        flag = true;
                        break;
                    }
                }
                else if(s.equals(")")||s.equals("]") && stack.isEmpty()){
                    flag = true;
                    break;
                }

            }
            if(flag == true || !stack.isEmpty()) System.out.println("no");
            else System.out.println("yes");
        }

    }
}
